server/game/cards/04.2-TL/UpholdingAuthority.js

```diff
@@ -46,8 +46,9 @@ class UpholdingAuthority extends ProvinceCard {
             context.game.addMessage('{0} chooses not to discard anything', context.player);
             return;
         }
-        gameAction.setTarget(cards.splice(0, choice));
-        this.game.addMessage('{0} chooses to discard {1} cop{2} of {3}', context.player, choice, (choice === 1 ? 'y' : 'ies'), cards[0]);
+        let targetCards = cards.splice(0, choice);
+        gameAction.setTarget(targetCards);
+        this.game.addMessage('{0} chooses to discard {1} cop{2} of {3}', context.player, choice, (choice === 1 ? 'y' : 'ies'), targetCards[0]);
```

Comment:
`splice` is removing the items from the original `cards` array so when you got to old line 50 `cards[0]` was undefined.  The chat message was always going to say "Player1 chooses to discard 3 copies of [undefined]".  I'd noticed this when playing myself.

Reply:
Yes you are right and I noticed this after fixing it, but it didn't have any impact on the functionality so I left it.  I will push a new commit though so that it is clearer.